teps/0026-interceptor-plugins.md

```diff
@@ -0,0 +1,510 @@
+and make progress.
+-->
+
+## Requirements
+
+1. Operators can install one or more built-in interceptors.
+2. Operators can remove any of the built-in interceptors.
+3. Users can configure Interceptors with custom parameters including references to Secrets.
+4. Interceptor authors can add new interceptors that have the same features as any built in interceptors, namely access to the entire request and any custom user parameters
+5. Any current Interceptor configuration is still usable in the new model without requiring any manual changes on part of the end user.
+
+
+## Proposal
+
+1. Interceptors run as separate processes that communicate with the EventListener over HTTP. The EventListener will send a request to the interceptor with a body that includes incoming event details (headers, body, url), as well as any parameters needed by the interceptor (filter expression, refs to secrets etc). The interceptor response body follows a predefined format that includes information on whether processing should continue or not, as well as any additional fields that the interceptor can add to the event body. This is in line with TEP-0022 which proposes a change to make the incoming event body immutable to improve auditability, and thus interceptors can only add extra fields instead of being able to modify the body or headers.
+ 
+2. Interceptors are configured or installed via an Interceptor CRD that describes an address for the Interceptor as well as  a list of valid parameters that the interceptor accepts.
+
+3. Built-in interceptors are still packaged as part of Triggers but run as separate processes instead of in the controller.
+
+4. Users configure interceptors in their Triggers and EventListeners. There will be a change in the syntax but we will support both the old and new syntax to allow for backwards compatibility:
+
+```yaml
+# OLD:
+interceptors:
+  - cel:
+    filter: "header.match("EventType", "push")
+
+# NEW:
+interceptors:
+  - type: CEL
+    params:
+    - name: "filter"
+      value: header.match("EventType", "push") 
+```
+
+
+5. For backwards compatibility, we will keep supporting the old WebhookInterceptor interface until beta. However, we will recommend users use the new Interceptor interface instead for writing any new Interceptors.
+
+See the [Design Details](#design-details) section below for more.
+
+
+### Notes/Constraints/Caveats
+
+1. Deployment Footprint and Scaling Concerns
+
+  The proposal suggests one interceptor deployment per cluster. This can cause scaling concerns. With one interceptor serving multiple EventListeners, it is possible that one EL that has many incoming events (or has many triggers with many interceptors) starts making too many requests to the interceptor. This could result in Listeners being bottle-necked on the interceptor processing and then failing to process their triggers (or building up a large queue of triggers/events to process).
```

Comment:
>   The proposal suggests one interceptor deployment per cluster.

The proposal would allow for more than one interceptor if you wanted tho, right? e.g. you could have many CEL interceptors, each with a separate `InterceptorConfiguration` if you wanted?

Reply:
in theory yes, you can run multiple CEL services and have multiple InterceptorConfiguration CRs to point to them.